pkgs/dart_tooling_mcp_server/lib/src/mixins/analyzer.dart

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:analyzer/dart/analysis/analysis_context_collection.dart';
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:dart_mcp/server.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p;
+import 'package:watcher/watcher.dart';
+
+/// Mix this in to any MCPServer to add support for analyzing Dart projects.
+///
+/// The MCPServer must already have the [ToolsSupport] mixin applied.
+base mixin DartAnalyzerSupport on ToolsSupport, LoggingSupport {
+  /// The analyzed contexts.
+  AnalysisContextCollection? _analysisContexts;
+
+  /// All active directory watcher streams.
+  ///
+  /// The watcher package doesn't let you close watchers, you just have to
+  /// stop listening to their streams instead, so we store the stream
+  /// subscriptions instead of the watchers themselves.
+  final List<StreamSubscription> _watchSubscriptions = [];
+
+  @override
+  FutureOr<InitializeResult> initialize(InitializeRequest request) {
+    if (request.capabilities.roots == null) {
+      throw StateError(
+          'This server requires the "roots" capability to be implemented.');
+    }
+    registerTool(analyzeFilesTool, _analyzeFiles);
+    initialized.then(_listenForRoots);
+    return super.initialize(request);
+  }
+
+  @override
+  Future<void> shutdown() async {
+    await super.shutdown();
+    await _analysisContexts?.dispose();
+    for (var subscription in _watchSubscriptions) {
+      unawaited(subscription.cancel());
+    }
+    _watchSubscriptions.clear();
+  }
+
+  /// Lists the roots, and listens for changes to them.
+  ///
+  /// Whenever new roots are found, creates a new [AnalysisContextCollection].
+  void _listenForRoots([void _]) async {
+    rootsListChanged!.listen((event) async {
+      unawaited(_analysisContexts?.dispose());
+      _analysisContexts = null;
+      _createAnalysisContext(await listRoots(ListRootsRequest()));
+    });
+    _createAnalysisContext(await listRoots(ListRootsRequest()));
+  }
+
+  /// Creates an analysis context from a list of roots.
+  //
+  // TODO: Better configuration for the DART_SDK location.
+  void _createAnalysisContext(ListRootsResult result) async {
+    final sdkPath = Platform.environment['DART_SDK'];
```

Comment:
can we use Platform.resolvedExecutable to find this?

Reply:
We can't because that doesn't work when AOT compiled (you just get the AOT script).

Reply:
(and we have to AOT compile because pub breaks JSON servers communicating over stdin/stdout with compiling messages) cc @sigurdm @jonasfj this has been discussed in the past right?

Reply:
It should not print any compilation messages if no terminal is attached.
If that is not the case, please file an issue.

Reply:
I don't know how this is gonna be deployed.
You could also consider compiling to a jit-snapshot via. `dart --snapshot=mcp.snapshot` and running with `dart mcp.snapshot`. No pub involved, and still knows the dart executable...

Reply:
I think we want to deploy it as a single executable anyways, that makes the most sense, it is mostly annoying for development.

I definitely was running into this but IIRC it was using the MCP inspector and I don't know exactly how it was spawning things? For instance if you were to pass `runInShell: true` when spawning from a dart program, would it still print to stdout?

Reply:
`runInShell: true` should not make it print to stdout.
You need to attach a pty to the process, something like https://github.com/TerminalStudio/pty for it to print.

(there might be bugs, but they should be fixable). The logic is here: https://github.com/dart-lang/pub/blob/639536aa78f7382b4a948669bfdc283f830b2ae9/lib/src/io.dart#L742

Reply:
There is an env-var for overriding whatever behavior you are otherwise getting...